src/backend/cuda/kernel/meanshift.hpp

```diff
@@ -12,200 +12,152 @@
-
-__forceinline__ __device__
-int lIdx(int x, int y,
-              int stride1, int stride0)
-{
-    return (y*stride1 + x*stride0);
-}
-
-__forceinline__ __device__
-int clamp(int f, int a, int b)
-{
-    return max(a, min(f, b));
-}
-
-template<typename T, int channels>
-inline __device__
-void load2ShrdMem(T * shrd, const T * in,
-                  int lx, int ly,
-                  int shrdStride, int schStride,
-                  int dim0, int dim1,
-                  int gx, int gy,
-                  int ichStride, int inStride1, int inStride0)
-{
-    int gx_  = clamp(gx, 0, dim0-1);
-    int gy_  = clamp(gy, 0, dim1-1);
-#pragma unroll
-    for(int ch=0; ch<channels; ++ch)
-        shrd[lIdx(lx, ly, shrdStride, 1)+ch*schStride] = in[lIdx(gx_, gy_, inStride1, inStride0)+ch*ichStride];
-}
-
-template<typename T, int channels>
+template<typename AccType, typename T, int channels>
 static __global__
-void meanshiftKernel(Param<T> out, CParam<T> in,
-                     float space_, int radius, float cvar,
-                     uint iter, int nBBS0, int nBBS1)
+void meanshiftKernel(Param<T> out, CParam<T> in, int radius, float cvar, uint iter,
+                     int nBBS0, int nBBS1)
 {
-    SharedMemory<T> shared;
-    T * shrdMem = shared.getPointer();
-
-    // calculate necessary offset and window parameters
-    const int padding     = 2*radius + 1;
-    const int shrdLen     = blockDim.x + padding;
-    const int schStride   = shrdLen*(blockDim.y + padding);
-    // the variable ichStride will only effect when we have >1
-    // channels. in the other cases, the expression in question
-    // will not use the variable
-    const int ichStride   = in.strides[2];
-
-    // gfor batch offsets
-    unsigned b2 = blockIdx.x / nBBS0;
-    unsigned b3 = blockIdx.y / nBBS1;
+    unsigned b2   = blockIdx.x / nBBS0;
+    unsigned b3   = blockIdx.y / nBBS1;
     const T* iptr = (const T *) in.ptr + (b2 *  in.strides[2] + b3 *  in.strides[3]);
     T*       optr = (T *      )out.ptr + (b2 * out.strides[2] + b3 * out.strides[3]);
+    const int gx  = blockDim.x * (blockIdx.x-b2*nBBS0) + threadIdx.x;
+    const int gy  = blockDim.y * (blockIdx.y-b3*nBBS1) + threadIdx.y;
 
-    const int lx = threadIdx.x;
-    const int ly = threadIdx.y;
-
-    const int gx = blockDim.x * (blockIdx.x-b2*nBBS0) + lx;
-    const int gy = blockDim.y * (blockIdx.y-b3*nBBS1) + ly;
+    if (gx>=in.dims[0] || gy>=in.dims[1])
+        return;
 
-    // pull image to local memory
-    for (int b=ly, gy2=gy; b<shrdLen; b+=blockDim.y, gy2+=blockDim.y) {
-        // move row_set get_local_size(1) along coloumns
-        for (int a=lx, gx2=gx; a<shrdLen; a+=blockDim.x, gx2+=blockDim.x) {
-            load2ShrdMem<T, channels>(shrdMem, iptr, a, b, shrdLen, schStride,
-                    in.dims[0], in.dims[1], gx2-radius, gy2-radius, ichStride,
-                    in.strides[1], in.strides[0]);
-        }
-    }
+    int i = gx;
+    int j = gy;
```

Comment:
Why is shared mem removed? Do you see better performance accessing global memory directly?